enhancements/network/configure-ovs-alternative.md

```diff
@@ -0,0 +1,562 @@
+advanced use cases, but it will require more complex configuration than
+basic deployments need. A possible followup would be to reimplement
+configure-ovs.sh in terms of this new mechanism so there is a simple path
+here too, but that is not required for the initial implementation.
+
+An example of a use case this is not intended to address would be a
+single-nic deployment on a cloud platform with no unusual network settings.
+In that case manual configuration of br-ex would be an unnecessary extra
+burden on the user and the logic to handle it automatically would be fairly
+simple so using configure-ovs is far less likely to cause problems.
+
+## Proposal
+
+This mechanism will be similar to the day-1 network configuration described
+in enhancements/network/baremetal-ipi-network-configuration.md. However,
+there will be some important differences, which will be discussed below.
+
+At a high level, we will provide an interface for the deployer to specify
+per-host interface configuration. Configuration for an OpenVSwitch bridge
+named br-ex will be mandatory. The format for the configuration will be
+NMState YAML, which will be provided to each node via Ignition. Around the
+same point in the boot process that configure-ovs would run, we will run
+NMState to apply the provided configuration. This will only be done on first
+boot because subsequent changes will be made using the Kubernetes-NMState
+operator.
+
+### Important differences from baremetal day-1
+
+Instead of pre-processing the NMState config into NetworkManager
+nmconnection files, we will simply write the NMState file directly to
+the host disk. As of 4.14 we have NMState available at the host level
+so it will no longer be necessary to run NMState in a container.
+
+Since this is intended to be a replacement for configure-ovs, it must
+work in a cross-platform manner. The baremetal feature is implemented in the
+baremetal-operator, which means non-baremetal platforms cannot use it.
+
+However, this is not a replacement for the baremetal feature (and any other
+platform-specific deploy-time network configuration tools). This is because
+minimal networking is required in order to pull Ignition, and since this
+configuration will be provided through Ignition it will still be necessary
+to have platform-specific methods of injecting network config before Ignition
+runs. However, the platform-specific part can be much simpler because it
+won't need to handle every possible network configuration, just enough to
+reach the Ignition endpoint. In some use cases, platform-specific configuration
+won't even be needed.
+
+### Workflow Description
+
+When the user is populating install-config they will provide the necessary
+configuration in the networking section (TODO: Is this the right place?).
+It will look something like this:
+
+networking:
+  networkType: OVNKubernetes
+  machineNetwork:
+  - cidr: 192.0.2.0/24
+  [...]
+  hostConfig:
+  - name: master-0
+    networkConfig:
```

Comment:
I suspect we would use the results of nodeip-configuration to generate the policy. I'm not sure exactly how that would work, but it's definitely something I'd like to look into. It's not my primary goal right now because I want to handle the advanced cases where configure-ovs struggles, but I do see nmpolicy as a potential way to completely replace configure-ovs.

On a more general note, a per-cluster config (maybe with the possibilty of overriding on a per-node basis) may be a solution to platforms where we don't have a way to map per-node configs ahead of time (like if the hostname is not predictable). I don't know how important this feature will be in those cases, but I have heard some discussion of needing to override the MTU at deploy time, so this might provide a solution to that. I think with my current machine-config prototype it would be pretty easy to extend to include a per-cluster config so maybe I'll give that a try and update the design if it works well.